indico/util/string.py

```diff
@@ -32,6 +36,92 @@
 from sqlalchemy import ForeignKeyConstraint, inspect
 
 
+class AutoLinkExtension(markdown.extensions.Extension):
+    def __init__(self, rules: list[list], **kwargs):
+        self.rules = rules
+        super().__init__(**kwargs)
+
+    # flake8: noqa: N802
+    def extendMarkdown(self, md: markdown.Markdown):
+        for n, rule in enumerate(self.rules):
+            md.inlinePatterns.register(AutoLinkInlineProcessor(rule['regex'], md, rule['url']), f'linker_{n}', 50)
+
+
+class AutoLinkInlineProcessor(markdown.inlinepatterns.InlineProcessor):
+
+    # exclude subsitution within links (nesting)
+    ANCESTOR_EXCLUDES = ('a',)
+
+    def __init__(self, pattern: str, md: markdown.Markdown, url: str):
+        self.url = url
+        super().__init__(pattern, md)
+
+    # flake8: noqa: N802
+    def handleMatch(self, m: re.Match, data: str):
+        el = Element('a')
+        el.set('href', self.url.format(m.group(0), *m.groups()))
+        el.text = m.group(0)
+        return el, m.start(0), m.end(0)
+
+
+class HTMLLinker:
+    """An HTML parser which applies autolinker rules."""
+
+    def __init__(self, rules: list[list]):
+        self.rules = rules
+
+    def _walk(self, tree_gen: t.Iterator[dict]):
+        can_parse = True
+        for item in tree_gen:
+            # we will ignore stuff inside links, to avoid nesting
+            if item['type'] in {'StartTag', 'EndTag'} and item['name'] in {'html', 'head', 'body'}:
+                pass
+            elif item['type'] == 'StartTag' and item['name'] == 'a':
+                can_parse = False
+                yield item
+            elif item['type'] == 'EndTag' and item['name'] == 'a':
+                can_parse = True
+                yield item
+            elif item['type'] == 'Characters' and can_parse:
+                text = item['data']
+
+                # "tokenize" text
+                tokens = []
+                last_idx = 0
+                for m in re.finditer(r'|'.join(r['regex'] for r in self.rules), text):
```

Comment:
Do we need `(?:...)` around the individual regexps or is `|` binding weak enough so it always does the right job even without it?

Reply:
Good point. I'm not sure, but it can't hurt...